lib/backend/buffer.go

```diff
@@ -36,16 +36,17 @@ import (
 type CircularBuffer struct {
 	sync.Mutex
 	*log.Entry
-	ctx      context.Context
-	cancel   context.CancelFunc
-	events   []Event
-	start    int
-	end      int
-	size     int
-	watchers *watcherTree
+	ctx          context.Context
+	cancel       context.CancelFunc
+	events       []Event
+	start        int
+	end          int
+	size         int
+	init, closed bool
```

Comment:
Is `closed` necessary? Wouldn't it suffice to check `ctx.Err()`?

Reply:
`closed` is necessary because the parent context might be canceled while the buffer lock is held, so using the context as the indicator of the closed state might cause inconsistent/racy behavior.

In thinking about your comment though, I've realized that the buffer doesn't really need to embed a context now that we're using `closed`, so I'm going to factor out the context.